fyrox-impl/src/scene/tilemap/edit.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2019-present Dmitry Stepanov and Fyrox Engine contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use crate::{
+    core::{
+        algebra::{Matrix4, Vector2, Vector3, Vector4},
+        parking_lot::Mutex,
+    },
+    scene::mesh::vertex::StaticVertex,
+};
+use fxhash::{FxHashMap, FxHashSet};
+use std::{fmt::Debug, sync::Arc};
+
+use super::*;
+
+/// A reference to TileMapEditorData. A TileMap keeps one of these if it is in the editor
+/// being edited so that the TileMap can render a partially-constructed state.
+pub type TileMapEditorDataRef = Arc<Mutex<TileMapEditorData>>;
+
+/// The state of a tile map that is being edited.
+#[derive(Default, Debug)]
+pub struct TileMapEditorData {

```

Comment:
It does not refer to anything specific to the Fyrox editor. It is a general-purpose facility that *any* editor might find useful, even perhaps an in-game level editor. It just changes how the tile map renders itself in ways that happen to be useful to the Fyrox editor.

Reply:
Certain important effects would be far more awkward to achieve if we could not control how the tile map is rendered and we are forced to work around a tile map that always renderers itself normally. It is especially important to be able to tell a tile map to not render tiles at certain positions without needing to actually remove those tiles. If we *must* remove all specialized rendering from the `TileMap` node, then I would suggest that we should create an editor-only version of the node that is added when the interaction mode is activated while the actual `TileMap` node is made invisible.

Reply:
Could you please rename it to something else that won't contain `editor`? Previous version contained `overlay_tiles` field which was used in the editor only, but didn't specifically tell that in the engine code.

Reply:
Changing this name has inspired me to also change how it works. I am working on a general-purpose system that games can use to change how the tile map renderers without changing the underlying data: making tiles invisible, temporarily replacing one tile with another, arbitrary rendering, and so on. It will do everything that the editor needs, plus games should be able to create effects beyond what the editor needs.

This also reminds me that *animated* tiles should be possible. I am unsure how that should work.